src/metrics_server.rs

```diff
@@ -0,0 +1,171 @@
+use hyper::{
+    header::CONTENT_TYPE,
+    service::{make_service_fn, service_fn},
+    Body, Method, Request, Response, Server,
+};
+use prometheus::{Encoder, Gauge, IntCounter, IntGauge, Opts};
+use std::{convert::Infallible, env};
+
+pub type Result<T, E> = std::result::Result<T, E>;
+use chrono::{DateTime, NaiveDateTime, Utc};
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+pub fn try_create_int_counter(name: &str, help: &str) -> Result<IntCounter, prometheus::Error> {
+    let opts = Opts::new(name, help);
+    let counter = IntCounter::with_opts(opts)?;
+    prometheus::register(Box::new(counter.clone()))?;
+    Ok(counter)
+}
+
+pub fn try_create_int_gauge(name: &str, help: &str) -> Result<IntGauge, prometheus::Error> {
+    let opts = Opts::new(name, help);
+    let gauge = IntGauge::with_opts(opts)?;
+    prometheus::register(Box::new(gauge.clone()))?;
+    Ok(gauge)
+}
+
+pub fn try_create_gauge(name: &str, help: &str) -> Result<Gauge, prometheus::Error> {
+    let opts = Opts::new(name, help);
+    let gauge = Gauge::with_opts(opts)?;
+    prometheus::register(Box::new(gauge.clone()))?;
+    Ok(gauge)
+}
+
+lazy_static! {
+    pub static ref BLOCK_PROCESSED_TOTAL: IntCounter =
+        try_create_int_counter("total_blocks_processed", "Total number of blocks processed")
+            .unwrap();
+    pub static ref LATEST_BLOCK_HEIGHT: IntGauge = try_create_int_gauge(
+        "latest_block_height",
+        "Latest Block Height Of Indexer Thus Far"
+    )
+    .unwrap();
+    pub static ref LATEST_BLOCK_TIMESTAMP: Gauge = try_create_gauge(
+        "latest_block_timestamp",
+        "Timestamp of when last block was indexed."
+    )
+    .unwrap();
+}
+
+enum ProbeResult {
+    IndexerStarted,
+    IndexerNotStarted,
+    SomethingIsWrong,
+}
+
+async fn serve_req(req: Request<Body>) -> Result<Response<Body>, hyper::Error> {
+    let response = match (req.method(), req.uri().path()) {
+        (&Method::GET, "/metrics") => {
+            let encoder = prometheus::TextEncoder::new();
+
+            let mut buffer = Vec::new();
+            if let Err(e) = encoder.encode(&prometheus::gather(), &mut buffer) {
+                eprintln!("could not encode custom metrics: {}", e);
```

Comment:
Good point, I think we should actually be a `tracing::error!` to catch this error. `eprintln!` came from the boilerplate code that sets ups Prometheus metrics, and this macro prints a formatted error to stdErr. 

Since we are using tracing for logging, we can use tracing::error! instead to be consistent.

Reply:
Please open the issue in all the projects where you've done this and create PRs to fix it when you have time